spark/src/main/java/org/apache/iceberg/spark/actions/BaseRemoveFilesSparkAction.java

```diff
@@ -0,0 +1,216 @@
+  private static final ExecutorService DEFAULT_DELETE_EXECUTOR_SERVICE = null;
+
+  private final Table table;
+  private final TableOperations ops;
+  private final Consumer<String> defaultDelete = new Consumer<String>() {
+    @Override
+    public void accept(String file) {
+      ops.io().deleteFile(file);
+    }
+  };
+
+  private Consumer<String> deleteFunc = defaultDelete;
+  private ExecutorService deleteExecutorService = DEFAULT_DELETE_EXECUTOR_SERVICE;
+
+  public BaseRemoveFilesSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    this.ops = ((HasTableOperations) table).operations();
+  }
+
+  @Override
+  protected RemoveFiles self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    JobGroupInfo info = newJobGroupInfo("REMOVE-FILES", "REMOVE-FILES");
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    boolean streamResults = PropertyUtil.propertyAsBoolean(options(), STREAM_RESULTS, false);
+    Dataset<Row> validFileDataset = buildValidFileDF(ops.current()).distinct();
+    RemoveFilesActionResult result;
+    if (streamResults) {
+      result = deleteFiles(validFileDataset.toLocalIterator());
+    } else {
+      result = deleteFiles(validFileDataset.collectAsList().iterator());
+    }
+    return result;
+  }
+
+  private Dataset<Row> appendTypeString(Dataset<Row> ds, String type) {
+    return ds.select(new Column("file_path"), functions.lit(type).as("file_type"));
+  }
+
+  private Dataset<Row> buildValidFileDF(TableMetadata metadata) {
+    Table staticTable = newStaticTable(metadata, this.table.io());
+    return appendTypeString(buildValidDataFileDF(staticTable), DATA_FILE)
+        .union(appendTypeString(buildManifestFileDF(staticTable), MANIFEST))
+        .union(appendTypeString(buildManifestListDF(staticTable), MANIFEST_LIST))
+        .union(appendTypeString(buildOtherMetadataFileDF(ops), OTHERS));
+  }
+
+  protected Dataset<Row> buildOtherMetadataFileDF(TableOperations tableOps) {
+    List<String> otherMetadataFiles = Lists.newArrayList();
+    otherMetadataFiles.add(tableOps.metadataFileLocation("version-hint.text"));
+
+    Set<String> metadataFiles = new HashSet<>();
+    TableMetadata metadata = tableOps.current();
+    String location = metadata.metadataFileLocation();
+    metadataFiles.add(location);
+    getMetadataFiles(location, metadataFiles, tableOps.io());
+    otherMetadataFiles.addAll(metadataFiles);
+    return spark().createDataset(otherMetadataFiles, Encoders.STRING()).toDF("file_path");
+  }
+
+  private void getMetadataFiles(String metadataFileLocation, Set<String> metaFiles, FileIO io) {
+    if (metadataFileLocation == null) {
+      return;
+    }
+    long minTimeStamp = Long.MAX_VALUE;
+    String minMetadataLocation = null;
+    TableMetadata metadata = TableMetadataParser.read(io, metadataFileLocation);
```

Comment:
Since this seems to be running recursively, is it possible that a previous metadata file's "previousFiles" is already cleaned up an no longer exist, and here when we read a non existing file it will throw exception?